### ISSUE TITLE
Macros don't expand correctly under VS2019, possibly because of old macro-utils-c dep missing later fixes

### ISSUE THREAD

**Issue by arsing** [Azure/azure-c-shared-utility#550]
(This issue spans multiple repos, specifically it affects every repo that has a dep on [macro-utils-c.](https://github.com/Azure/macro-utils-c) I'm reporting it here because this is one of our deps that has this issue, the other being [azure-utpm-c.](https://github.com/Azure/azure-utpm-c))

We at [Azure/iotedge](https://github.com/Azure/iotedge) have [a submodule dep](https://github.com/Azure/iotedge/tree/release/1.1/edgelet/hsm-sys/azure-iot-hsm-c/deps) on this repo for our code, currently at the latest [2020-12-09](https://github.com/Azure/azure-c-shared-utility/releases/tag/2020-12-09) release. Our code that uses ctest.h, gballoc.h, etc doesn't compile on Windows with VS 2019 because some macros are not expanded correctly.

For example, [this line in ctest.h](https://github.com/Azure/ctest/blob/4aa84c45236ecda21983fb682f542f17d0b21e19/inc/ctest.h#L82) expands to:

```
typedef enum CTEST_FUNCTION_TYPE_TAG { MU_FOR_EACH_1_8 ( MU_DEFINE_ENUMERATION_CONSTANT, CTEST_FUNCTION_TYPE_INVALID, CTEST_BEGIN_SUITE, CTEST_END_SUITE, CTEST_TEST_FUNCTION, CTEST_TEST_SUITE_INITIALIZE, CTEST_TEST_SUITE_CLEANUP, CTEST_TEST_FUNCTION_INITIALIZE, CTEST_TEST_FUNCTION_CLEANUP)} CTEST_FUNCTION_TYPE; ...
```

Notice the `MU_FOR_EACH_1_8` macro has not been expanded. This causes weird syntax errors.

I discovered that if I change [this code in macro-utils-c](https://github.com/Azure/macro-utils-c/blob/5926caf4e42e98e730e6d03395788205649a3ada/inc/azure_macro_utils/macro_utils_generated.h#L9054-L9058) such that the non-MSVC definition is used for MSVC, ie the definition that uses `(` instead of `LPAREN`, the macro started expanding correctly. So whatever hack the using `LPAREN` instead of `(` was solving for MSVC, it's apparently now harmful for MSVC 2019.

Based on this, I realized that the macro-utils-c repo has actually changed this code to use the original non-MSVC definition on MSVC 2019, specifically with these two commits:

- https://github.com/Azure/macro-utils-c/commit/4240222d5db6d81fa0d211994797c1acb767cd6f#diff-159abdbc930962258f6b59e4328759cd6b01b9a04f293801a6ec9539c0c77a0eR123

- https://github.com/Azure/macro-utils-c/commit/4cbcec60757b904e47bff4aa5a2e5addae5bcf57#diff-bfc53cce44d87ebab4cb99fa5b055549a05b8a12da7b8838cdd61f622b5ce3a5R9072

Also, our code used to compile on VS 2019 just fine at some point in the past but was only recently discovered to be broken. You guys probably have more context (cc @anporumb), but I guess at some point MSVC 2019 fixed whatever bug required the `LPAREN` hack which is why it's only broken for us now.

There are other macros that also don't expand correctly because they also use `LPAREN` under MSVC. Also, our dep tree from azure-c-shared-utility and azure-utpm-c pulls in multiple copies of macro-utils-c, so we can't ad-hoc patch them all ourselves. So if you confirm my hypothesis that (at least) those two commits are needed to compile for MSVC 2019, I'd appreciate if you could update all the Azure repos that depend on macro-utils-c to a commit that has all the necessary fixes.

**Comment by ericwol-msft** [comment]
I have no problems building the azure-c-shared-utility repo. I'm using Microsoft Visual Studio Enterprise 2019 Version 16.11.2

```
124>consolelogger.c
124>Generating Code...
124>schemalib_ut_exe.vcxproj -> C:\Repos\azure-iot-sdk-c(longhaul)\cmake\serializer\tests\schemalib_ut\Debug\schemalib_ut_exe.exe
125>------ Build started: Project: ALL_BUILD, Configuration: Debug x64 ------
125>Building Custom Rule C:/Repos/azure-iot-sdk-c(longhaul)/CMakeLists.txt
126>------ Skipped Build: Project: INSTALL, Configuration: Debug x64 ------
126>Project not selected to build for this solution configuration 
========== Build: 120 succeeded, 0 failed, 0 up-to-date, 6 skipped ==========
```

**Comment by arsing** [comment]
Does your schemalib_ut import ctest.h ? If so, what does the preprocessed file say about the expansion of the line I mentioned?

**Comment by ericwol-msft** [comment]
There is no schemalib_ut in Azure/azure-c-shared-utility

**Comment by arsing** [comment]
Yes. Because it's in azure-iot-sdk-c. Which is what you built and showed me a successful build of.

**Comment by ericwol-msft** [comment]
Well, you are posting in Azure/azure-c-shared-utility, so that is what I'm building. Why are you posting Azure/azure-iot-sdk-c issues in Azure/azure-c-shared-utility???

**Comment by arsing** [comment]
>Well, you are posting in Azure/azure-c-shared-utility, so that is what I'm building.

Your build output contains:

>124>schemalib_ut_exe.vcxproj -> C:\Repos\azure-iot-sdk-c(longhaul)\cmake\serializer\tests\schemalib_ut\Debug\schemalib_ut_exe.exe

so it doesn't seem to me that you're building just azure-c-shared-utility.

>Why are you posting Azure/azure-iot-sdk-c issues in Azure/azure-c-shared-utility???

- I explained why I posted here in the OP.

- You're the one who brought up azure-iot-sdk-c by posting your build output of it. I didn't say anything about it. Why does that make my issue an azure-iot-sdk-c issue?

**Comment by ericwol-msft** [comment]
No. Azure/azure-c-shared-utility does not directly include ctest.h

**Comment by arsing** [comment]
Okay. Then your successful build isn't relevant. If you want to see the issue for yourself you'll need to write some code that imports that header.

**Comment by ericwol-msft** [comment]
So you have no issues building the Azure/azure-c-shared-utility SDK?

**Comment by ericwol-msft** [comment]
I have no problems building this repo and all the gates are green. Please provide detailed repo steps.

**Comment by arsing** [comment]
You do have tests that include gballoc.h, eg azure_base32_ut, but it compiles successfully. And of course its preprocessed input shows that the `MU_FOR_EACH_1` macro is being expanded correctly.

I also confirmed that it only compiles successfully with the `LPAREN` version of the `MU_FOR_EACH_1` macro and not the `(` version. Which is to say, replicating the changes from the two macro-utils-c commits actually breaks the build.

I also compared the cl.exe commandlines and they're using identical flags. Windows SDK version in the vcxproj is identical too.

@anporumb Do you know what's going on? Are your two commits that I linked in the OP supposed to be required for MSVC 2019? If so, why would MSVC 2019 seemingly require the `(` version in one project but the `LPAREN` version in another?

**Comment by ericwol-msft** [comment]
I'm closing this issue. There is no SLA around unit tests and their primary purpose is for internal build verification. Currently our gates are passing and there is no plan to change any of this.

**Comment by arsing** [comment]
We aren't building the project unit tests in the first place. We're building *our* repo's unit tests. I explained this in https://github.com/Azure/azure-c-shared-utility/issues/535#issuecomment-912045533

**Comment by ericwol-msft** [comment]
I did not see any link to your repo, so I assumed you were referring to Azure/azure-c-shared-utility. Sorry about that.

**Comment by arsing** [comment]
I mentioned it in the first paragraph of the OP.

**Comment by ericwol-msft** [comment]
Once again, sorry about that.

**Comment by arsing** [comment]
So, you're still not going to reopen it?

**Comment by ericwol-msft** [comment]
My understanding is that you are having tool chain issues when building an external dependent repo. You posted in Azure/azure-c-shared-utility because we also have an external dependency on this repo, due to our usage in our unit tests. We are not currently experiencing any failures, most likely because we have a dependency on a different version of the external repo. Once again, I'm deeply sorry you are having issues with your tool chain and I thank you for reaching out to us. Currently there is nothing actionable for us on this topic nor any plans to change or update our unit test framework.

**Comment by arsing** [comment]
>We are not currently experiencing any failures, most likely because we have a dependency on a different version of the external repo.

If you had a different version of the dependency than us, I wouldn't have opened the issue here. We don't have any dependency on umock-c. I opened the issue here because the issue arises from *your* use of umock-c, and the issue asks you to update *your* umock-c dependency to a version that supposedly has the fix (the two commits I linked to).

Anyway, I'm not going to waste any more time on this. Have a good day.
